src/core/service.c

```diff
@@ -4908,6 +4911,24 @@ static void service_handoff_timestamp(
         unit_add_to_dbus_queue(u);
 }
 
+static void service_notify_pidref(Unit *u, PidRef *parent_pidref, PidRef *child_pidref) {
+        Service *s = ASSERT_PTR(SERVICE(u));
+
+        assert(parent_pidref);
+        assert(child_pidref);
+
+        if (pidref_equal(&s->main_pid, parent_pidref)) {
+                pidref_done(&s->main_pid);
+                s->main_pid = TAKE_PIDREF(*child_pidref);
```

Comment:
service_set_main_pidref?

Reply:
@YHNdnzj I think the problem is service_set_main_pidref makes an assumption the main PID will be our child, otherwise it logs and sets main_pid_alien: https://github.com/systemd/systemd/blob/89099136d7c28310d7159817c90e970c62a995ec/src/core/service.c#L245

In PID namespaces, main PID won't be our child but its still running systemd-executor. So I don't think we want to set main_pid_alien

The rest of the function looks useful e.g. adjusting starting timestamp. Maybe I will refactor to use part of this function.

Reply:
Actually...since its not our child, I think we do want to set main_pid_alien = true since we won't get SIGCHLD if it exits. So lets call this function as is